bluesky/callbacks/tiled_writer.py

```diff
@@ -0,0 +1,206 @@
+from tiled.client import from_profile, from_uri
+from tiled.structures.array import ArrayStructure, BuiltinDtype
+from tiled.structures.core import Spec, StructureFamily
+from tiled.structures.data_source import Asset, DataSource, Management
+from tiled.structures.table import TableStructure
+
+MIMETYPE_LOOKUP = {
+    "hdf5": "application/x-hdf5",
+    "ADHDF5_SWMR_STREAM": "application/x-hdf5",
+    "AD_HDF5_SWMR_SLICE": "application/x-hdf5",
+}
+
+
+class TiledWriter:
+    "Write metadata and data from Bluesky documents into Tiled."
+
+    def __init__(self, client):
+        self.client = client
+        self._run_router = RunRouter([self._factory])
+
+    def _factory(self, name, doc):
+        return [_RunWriter(self.client)], []
+
+    @classmethod
+    def from_uri(cls, uri, **kwargs):
+        client = from_uri(uri, **kwargs)
+        return cls(client)
+
+    @classmethod
+    def from_profile(cls, profile, **kwargs):
+        client = from_profile(profile, **kwargs)
+        return cls(client)
+
+    def __call__(self, name, doc):
+        self._run_router(name, doc)
+
+
+class _RunWriter(DocumentRouter):
+    "Write the document from one Bluesky Run into Tiled."
+
+    def __init__(self, client):
+        self.client = client
+        self.start_stop_node = None
+        self._descriptor_nodes = {}  # references to descriptor containers by uid's
+        self._sr_nodes = {}
+        self._sr_cache = {}
+
+    def start(self, doc):
+        self.start_stop_node = self.client.create_container(
+            key=doc["uid"],
+            metadata={"start": doc},
+            specs=[Spec("BlueskyRun", version="1.0")],
+        )
+
+    def stop(self, doc):
+        doc = dict(doc)
+        doc.pop("start", None)
+        metadata = dict(self.start_stop_node.metadata) | {"stop": doc}
+        self.start_stop_node.update_metadata(metadata=metadata)
+
+    def descriptor(self, doc):
+        desc_name = doc["name"]
+        metadata = dict(doc)
+
+        # Extract the variable fileds of the descriptor that can change during the run
+        uid = metadata.pop("uid")
+        conf_dict = {uid: metadata.pop("configuration", {})}
+        time_dict = {uid: metadata.pop("time", None)}
+        var_fields = {"configuration": conf_dict, "time": time_dict}
+
+        # Get or create a descriptor container
+        if desc_name in self.start_stop_node.keys():
+            desc_node = self.start_stop_nodes[desc_name]
+        else:
+            desc_node = self.start_stop_node.create_container(
+                key=desc_name, metadata=metadata
+            )
+            desc_node.create_container(key="external")
+            desc_node.create_container(key="internal")
+
+        # Update the metadata with the variable fields
+        metadata = deep_update(dict(desc_node.metadata), var_fields)
+        desc_node.update_metadata(metadata)
+        self._descriptor_nodes[uid] = desc_node  # Keep a reference to the descriptor
```

Comment:
Should this move into the `else` block, for clarity? In the other case I think it is or should be a no-op.

Reply:
Same as above. The idea was that the `else` block would run first, create the Descriptor node, and populate it with (fixed) metadata. Then, the variable part of the metadata will be added after the condition.
Next time, when the descriptor is updated, there's no need to rewrite the fixed metadata (hence we only pull the node in the direct path of the condition `desc_node = self.root_nodes[desc_name]`, but the after the entire `if .. else` block the variable part is added like before).
Again, probably worth expanding this part of the code a bit more; i'll look into this.